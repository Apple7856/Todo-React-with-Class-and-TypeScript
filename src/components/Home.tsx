import { Button, Card, CardActionArea, CardActions, CardContent, FormControl, Grid, InputLabel, MenuItem, Select, TextField, Typography } from '@material-ui/core';
import React from 'react';
import { itemType } from '../interface';
import './createtodo.css';
import './cardbox.css';
import './home.css';

interface props {

}

interface state {
    getTitle: string,
    getDesc: string | number,
    getStatus: string,
    myNewInputData: itemType,
    items: itemType[],
    editItem: itemType,
    editForm: boolean,
    editId: string,
    checkStatus: boolean
}

class Home extends React.Component<props, state> {
    constructor(props: props) {
        super(props);
        this.state = {
            getTitle: "",
            getDesc: "",
            getStatus: "",
            myNewInputData: {},
            items: [],
            editItem: {},
            editForm: false,
            editId: "",
            checkStatus: false,
        }
    }

    addTodo() {
        if (this.state.getTitle && this.state.getDesc && this.state.getStatus) {
            console.log("Data Value");
            this.setState({
                myNewInputData: {
                    id: new Date().getTime().toString(),
                    title: this.state.getTitle,
                    desc: this.state.getDesc,
                    status: this.state.getStatus
                }
            })
        }
        this.setState({ getTitle: "" });
        this.setState({ getDesc: "" });
        this.setState({ getStatus: "" });
    }

    componentDidMount() {
        if (localStorage.getItem("data")) {
            this.setState({ items: JSON.parse(localStorage.getItem("data") || "") });
        }
    }

    componentDidUpdate(prevProps: props, prevState: state) {
        if (prevState.myNewInputData !== this.state.myNewInputData) {
            this.setState({ items: [...this.state.items, this.state.myNewInputData] })
        }
        if (prevState.items !== this.state.items) {
            localStorage.setItem("data", JSON.stringify(this.state.items));
        }
        if (prevState.editItem !== this.state.editItem) {
            this.setState({ getTitle: this.state.editItem.title || "" });
            this.setState({ getDesc: this.state.editItem.desc || "" });
            this.setState({ getStatus: this.state.editItem.status || "" });
            this.setState({ editId: this.state.editItem.id || "" });
            if (this.state.editItem.status === "Complete") {
                this.setState({ checkStatus: true });
            }
        }
    }

    editItem(id: string) {
        this.setState({ editForm: true });
        this.state.items.find((elem) => {
            return elem.id === id ? this.setState({ editItem: elem }) : ""
        })
    }

    deleteItem(id: string) {
        this.setState({
            items: this.state.items.filter((elem) => {
                return elem.id !== id
            })
        })
    }

    updateTodo() {
        this.setState({
            items: this.state.items.map((elem) => {
                if (elem.id === this.state.editId) {
                    return { ...elem, id: this.state.editId, title: this.state.getTitle, desc: this.state.getDesc, status: this.state.getStatus };
                } else {
                    return elem;
                }
            })
        })
        this.setState({ getTitle: "" });
        this.setState({ getDesc: "" });
        this.setState({ getStatus: "" });
        this.setState({ editId: "" });
        this.setState({ editItem: {} });
        this.setState({ editForm: false });
        this.setState({ checkStatus: false });
    }

    render() {
        return (
            <>
                <Grid container data-testid="home-1">
                    <Grid item xs={8} className="grid1">
                        <div className="cTodoBox">
                            <Typography variant='h6' className="text" data-testid="heading">
                                {
                                    this.state.editForm ? "Edit Todo" : "Create Todo"
                                }
                            </Typography>
                            <form className="form" autoComplete='off'>
                                <div className="titleBox">
                                    <TextField id="standard-basic" name="title" className="title" label="title" value={this.state.getTitle} onChange={(e) => this.setState({ getTitle: e.target.value.replace(/[^a-zA-Z]/ig, '') })} />
                                </div>
                                <div className="multilineBox">
                                    <TextField
                                        id="outlined-multiline-static"
                                        label="Desc"
                                        name="desc"
                                        className="multiline"
                                        multiline
                                        rows={4}
                                        variant="outlined"
                                        value={this.state.getDesc}
                                        onChange={(e) => this.setState({ getDesc: e.target.value })}
                                    />
                                </div>
                                <div className="formControlBox">
                                    <FormControl className="formControl">
                                        <InputLabel id="demo-simple-select-label">Status</InputLabel>
                                        <Select
                                            labelId="demo-simple-select-label"
                                            id="demo-simple-select"
                                            name="status"
                                            disabled = {this.state.checkStatus}
                                            value={this.state.getStatus}
                                            onChange={(e: any) => this.setState({ getStatus: e.target.value })}
                                        >
                                            <MenuItem value="Todo">Todo</MenuItem>
                                            <MenuItem value="InProgress">InProgress</MenuItem>
                                            <MenuItem value="Complete">Complete</MenuItem>
                                        </Select>
                                    </FormControl>
                                </div>
                                <div className="button">
                                    <Button variant="contained" color="primary" onClick={this.state.editForm ? () => this.updateTodo() : () => this.addTodo()} data-testid="button" className='btn'>
                                        {
                                            this.state.editForm ? "Update Todo" : "Add Todo"
                                        }
                                    </Button>
                                </div>
                            </form>
                        </div>

                    </Grid>
                    <Grid item xs={4} className="grid2">
                        <div>
                            {
                                this.state.items.map((elem, i) => {
                                    {
                                        return (<><Card className="card" key={elem.id}>
                                            <CardActionArea>
                                                <CardContent>
                                                    <Typography gutterBottom variant='h5' color="textSecondary" component="h2">{elem.title}</Typography>
                                                    <div className="desc">
                                                        <Typography variant="body2" color="textSecondary" component="p">
                                                            {elem.desc}
                                                        </Typography>
                                                    </div>
                                                    <Typography variant="body2" color="textSecondary" className="status" component="p">
                                                        Status: <span> {elem.status}</span>
                                                    </Typography>
                                                </CardContent>
                                            </CardActionArea>
                                            <CardActions>
                                                <Button size="small" color="primary" onClick={() => this.editItem(elem.id || "")}>
                                                    Edit
                                                </Button>
                                                <Button size="small" color="secondary" onClick={() => this.deleteItem(elem.id || "")}>
                                                    Delete
                                                </Button>
                                            </CardActions>
                                        </Card>
                                        </>)
                                    }
                                })
                            }
                        </div>
                    </Grid>
                </Grid>
            </>
        )
    }
}

export default Home;

