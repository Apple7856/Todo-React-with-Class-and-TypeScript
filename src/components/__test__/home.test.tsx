import React from 'react';
import { render, cleanup, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Home from '../Home';


describe('Home page testing', () => {

    test("test", () => {
        expect(true).toBe(true);
    });

    test('heading Testing', () => {
        const component = render(<Home />);
        const heading = component.getByTestId('heading');
        expect(heading.textContent).toBe("Create Todo");
    })

    test('Button Testing', () => {
        const component = render(<Home />);
        const button = component.getByTestId('button');
        expect(button.textContent).toBe('Add Todo');
    })

    test('snapShot testing', () => {
        const component = render(<Home />);
        expect(component).toMatchSnapshot();
    })
})

