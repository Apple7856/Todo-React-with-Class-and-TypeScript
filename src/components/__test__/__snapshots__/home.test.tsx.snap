// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page testing snapShot testing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiGrid-root MuiGrid-container"
        data-testid="home-1"
      >
        <div
          class="MuiGrid-root grid1 MuiGrid-item MuiGrid-grid-xs-8"
        >
          <div
            class="cTodoBox"
          >
            <h6
              class="MuiTypography-root text MuiTypography-h6"
              data-testid="heading"
            >
              Create Todo
            </h6>
            <form
              autocomplete="off"
              class="form"
            >
              <div
                class="titleBox"
              >
                <div
                  class="MuiFormControl-root MuiTextField-root title"
                >
                  <label
                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
                    data-shrink="false"
                    for="standard-basic"
                    id="standard-basic-label"
                  >
                    title
                  </label>
                  <div
                    class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
                  >
                    <input
                      aria-invalid="false"
                      class="MuiInputBase-input MuiInput-input"
                      id="standard-basic"
                      name="title"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </div>
              <div
                class="multilineBox"
              >
                <div
                  class="MuiFormControl-root MuiTextField-root multiline"
                >
                  <label
                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
                    data-shrink="false"
                    for="outlined-multiline-static"
                    id="outlined-multiline-static-label"
                  >
                    Desc
                  </label>
                  <div
                    class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-multiline MuiOutlinedInput-multiline"
                  >
                    <textarea
                      aria-invalid="false"
                      class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline"
                      id="outlined-multiline-static"
                      name="desc"
                      rows="4"
                    />
                    <fieldset
                      aria-hidden="true"
                      class="PrivateNotchedOutline-root-9 MuiOutlinedInput-notchedOutline"
                    >
                      <legend
                        class="PrivateNotchedOutline-legendLabelled-11"
                      >
                        <span>
                          Desc
                        </span>
                      </legend>
                    </fieldset>
                  </div>
                </div>
              </div>
              <div
                class="formControlBox"
              >
                <div
                  class="MuiFormControl-root formControl"
                >
                  <label
                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
                    data-shrink="false"
                    id="demo-simple-select-label"
                  >
                    Status
                  </label>
                  <div
                    class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
                  >
                    <div
                      aria-haspopup="listbox"
                      aria-labelledby="demo-simple-select-label demo-simple-select"
                      class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
                      id="demo-simple-select"
                      role="button"
                      tabindex="0"
                    >
                      <span>
                        â€‹
                      </span>
                    </div>
                    <input
                      aria-hidden="true"
                      class="MuiSelect-nativeInput"
                      name="status"
                      tabindex="-1"
                      value=""
                    />
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSelect-icon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M7 10l5 5 5-5z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
              <div
                class="button"
              >
                <button
                  class="MuiButtonBase-root MuiButton-root MuiButton-contained btn MuiButton-containedPrimary"
                  data-testid="button"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiButton-label"
                  >
                    Add Todo
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </div>
            </form>
          </div>
        </div>
        <div
          class="MuiGrid-root grid2 MuiGrid-item MuiGrid-grid-xs-4"
        >
          <div />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiGrid-root MuiGrid-container"
      data-testid="home-1"
    >
      <div
        class="MuiGrid-root grid1 MuiGrid-item MuiGrid-grid-xs-8"
      >
        <div
          class="cTodoBox"
        >
          <h6
            class="MuiTypography-root text MuiTypography-h6"
            data-testid="heading"
          >
            Create Todo
          </h6>
          <form
            autocomplete="off"
            class="form"
          >
            <div
              class="titleBox"
            >
              <div
                class="MuiFormControl-root MuiTextField-root title"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
                  data-shrink="false"
                  for="standard-basic"
                  id="standard-basic-label"
                >
                  title
                </label>
                <div
                  class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiInput-input"
                    id="standard-basic"
                    name="title"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
            <div
              class="multilineBox"
            >
              <div
                class="MuiFormControl-root MuiTextField-root multiline"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
                  data-shrink="false"
                  for="outlined-multiline-static"
                  id="outlined-multiline-static-label"
                >
                  Desc
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-multiline MuiOutlinedInput-multiline"
                >
                  <textarea
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline"
                    id="outlined-multiline-static"
                    name="desc"
                    rows="4"
                  />
                  <fieldset
                    aria-hidden="true"
                    class="PrivateNotchedOutline-root-9 MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="PrivateNotchedOutline-legendLabelled-11"
                    >
                      <span>
                        Desc
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <div
              class="formControlBox"
            >
              <div
                class="MuiFormControl-root formControl"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
                  data-shrink="false"
                  id="demo-simple-select-label"
                >
                  Status
                </label>
                <div
                  class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
                >
                  <div
                    aria-haspopup="listbox"
                    aria-labelledby="demo-simple-select-label demo-simple-select"
                    class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
                    id="demo-simple-select"
                    role="button"
                    tabindex="0"
                  >
                    <span>
                      â€‹
                    </span>
                  </div>
                  <input
                    aria-hidden="true"
                    class="MuiSelect-nativeInput"
                    name="status"
                    tabindex="-1"
                    value=""
                  />
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSelect-icon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M7 10l5 5 5-5z"
                    />
                  </svg>
                </div>
              </div>
            </div>
            <div
              class="button"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained btn MuiButton-containedPrimary"
                data-testid="button"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiButton-label"
                >
                  Add Todo
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </form>
        </div>
      </div>
      <div
        class="MuiGrid-root grid2 MuiGrid-item MuiGrid-grid-xs-4"
      >
        <div />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
